app/src/main/java/com/keylesspalace/tusky/components/conversation/ConversationsFragment.kt

```diff
@@ -52,93 +58,128 @@ class ConversationsFragment : SFragment(), StatusActionListener, Injectable, Res
+            useAbsoluteTime = preferences.getBoolean("absoluteTimeView", false),
+            showBotOverlay = preferences.getBoolean("showBotOverlay", true),
+            useBlurhash = preferences.getBoolean("useBlurhash", true),
+            cardViewMode = CardViewMode.NONE,
+            confirmReblogs = preferences.getBoolean("confirmReblogs", true),
+            hideStats = preferences.getBoolean(PrefKeys.WELLBEING_HIDE_STATS_POSTS, false),
+            animateEmojis = preferences.getBoolean(PrefKeys.ANIMATE_CUSTOM_EMOJIS, false)
         )
 
-        adapter = ConversationAdapter(statusDisplayOptions, this, ::onTopLoaded, viewModel::retry)
+        adapter = ConversationAdapter(statusDisplayOptions, this)
+        loadStateAdapter = ConversationLoadStateAdapter(adapter::retry)
 
         binding.recyclerView.addItemDecoration(DividerItemDecoration(view.context, DividerItemDecoration.VERTICAL))
         layoutManager = LinearLayoutManager(view.context)
         binding.recyclerView.layoutManager = layoutManager
-        binding.recyclerView.adapter = adapter
+        binding.recyclerView.adapter = adapter.withLoadStateFooter(loadStateAdapter)
         (binding.recyclerView.itemAnimator as SimpleItemAnimator).supportsChangeAnimations = false
 
         binding.progressBar.hide()
         binding.statusView.hide()
 
         initSwipeToRefresh()
 
-        viewModel.conversations.observe(viewLifecycleOwner) {
-            adapter.submitList(it)
-        }
-        viewModel.networkState.observe(viewLifecycleOwner) {
-            adapter.setNetworkState(it)
+        lifecycleScope.launch {
+            viewModel.conversationFlow.collectLatest { pagingData ->
+                adapter.submitData(pagingData)
+            }
         }
 
-        viewModel.load()
-
+        adapter.addLoadStateListener { loadStates ->
+
+            loadStates.refresh.let { refreshState ->
+                if (refreshState is LoadState.Error) {
+                    binding.statusView.show()
+                    if (refreshState.error is IOException) {
+                        binding.statusView.setup(R.drawable.elephant_offline, R.string.error_network) {
+                            adapter.refresh()
+                        }
+                    } else {
+                        binding.statusView.setup(R.drawable.elephant_error, R.string.error_generic) {
+                            adapter.refresh()
+                        }
+                    }
+                } else {
+                    binding.statusView.hide()
+                }
+
+                if (refreshState ==  LoadState.Loading && adapter.itemCount == 0) {
+                    binding.recyclerView.hide()
+                    binding.progressBar.show()
+                } else {
+                    binding.recyclerView.show()
+                    binding.progressBar.hide()
+                }
+
+                if (refreshState is LoadState.NotLoading && !initialRefreshDone) {
```

Comment:
is `initialRefreshDone` ever set?

Reply:
good point, that is why it kept randomly jumping up 😅